homeassistant/components/bitvavo/sensor.py

```diff
@@ -0,0 +1,279 @@
+        return self._get_data_property("symbol")
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend."""
+        return CURRENCY_ICONS.get(self.unit_of_measurement, DEFAULT_COIN_ICON)
+
+    @property
+    def device_state_attributes(self):
+        """Return additional sensor state attributes."""
+        return {
+            "available": self._get_data_property("available"),
+            "in order": self._get_data_property("inOrder"),
+            f"{ASSET_VALUE_BASE}_value".lower(): float(
+                self._get_data_property("asset_value_in_base_asset")
+            ),
+            "unit_of_measurement": self._get_data_property("symbol"),
+            "source": "Bitvavo",
+        }
+
+
+class Order(CoordinatorEntity):
+    """Implementation of the order sensor."""
+
+    def __init__(self, coordinator, order):
+        """Initialize the sensor."""
+        super().__init__(coordinator)
+        self._order = order
+        self._icon = "mdi:checkbook"
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend."""
+        return self._icon
+
+    @property
+    def device_state_attributes(self):
+        """Return additional sensor state attributes."""
+        return {
+            "source": "Bitvavo",
+        }
+
+
+class OpenOrder(Order):
+    """Implementation of the open order sensor."""
+
+    def __init__(self, coordinator, order):
+        """Initialize the sensor."""
+        super().__init__(coordinator, order)
+
+        self._name = "Bitvavo Orders - Open"
+        self._unique_id = "bitvavo_orders_open"
+
+    def _get_data(self):
+        """Return the data from self.coordinator.data."""
+        return self.coordinator.data[CONF_OPEN_ORDERS]
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return len(self._get_data())
+
+    @property
+    def unique_id(self):
+        """Return a unique id for the sensor."""
+        return self._unique_id
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit the value is expressed in."""
+        return "Orders"
+
+
+class TotalAssetValue(CoordinatorEntity):
+    """Implementation of the total asset value sensor."""
+
+    def __init__(self, coordinator, currency):
+        """Initialize the sensor."""
+        super().__init__(coordinator)
+        self._currency = currency
+
+        self._name = f"Bitvavo Total Asset Value - {self._currency.upper()}"
+        self._unique_id = f"bitvavo_total_asset_value_{self._currency.lower()}"
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        total_base_asset_value = 0.0
+
+        for i in self.coordinator.data[CONF_BALANCES].keys():
```

Comment:
Let's calculate this all only once each update (pre-calculate).

Reply:
I moved the general base asset data calculation code to the coordinator update. The leftover state calculate code is per sensor.